src/datasets/features.py

```diff
@@ -33,18 +34,60 @@
 
 
 def string_to_arrow(type_str: str) -> pa.DataType:
-    if type_str not in pa.__dict__:
+    """
+    string_to_arrow takes a string pyarrow dtype and converts it to a pyarrow.DataType.
+
+    In effect, `dt == string_to_arrow(str(dt))`
+
+    This is necessary because the datasets.Value() primitive type is constructed using a string dtype, oftentimes
+
+    Value(dtype=str(pa_type))
+
+    But Features.type (via `get_nested_type()` expects to resolve Features into a pyarrow Schema,
+        which means that each Value() must resolve into its corresponding pyarrow.DataType, which is the purpose
+        of this function.
+    """
+    timestamp_regex = re.compile(r"^timestamp\[(.*)\]$")
+    timestamp_matches = timestamp_regex.search(type_str)
+    if timestamp_matches:
+        """
+        >>> import pyarrow as pa
+        >>> str(pa.timestamp('us'))
+        'timestamp[us]'
+        >>> str(pa.timestamp('us', tz='America/New_York'))
+        'timestamp[us, tz=America/New_York]'
+        """
+        timestamp_internals = timestamp_matches.group(1)
+        internals_regex = re.compile(r"^(s|ms|us|ns),\s*tz=([a-zA-Z0-9/_+:]*)$")
+        internals_matches = internals_regex.search(timestamp_internals)
+        if timestamp_internals in ["s", "ms", "us", "ns"]:
+            return pa.timestamp(timestamp_internals)
```

Comment:
I opted for multiple returns, but if you don't like this style, I'm happy to refactor this into assigning a `return_datatype` value and having a single return at the end.

Reply:
I'm fine with this, thanks :)